mdf_matio/validator.py

```diff
@@ -0,0 +1,362 @@
+    record_entries = []
+    for r in records:
+        record_entry = vald_gen.send(r)
+        record.entries.append(record_entry)
+    vald_gen.send(None)
+    ```
+    """
+    def __init__(self, schema_branch="master"):
+        """Create an MDFValidator.
+
+        Arguments:
+            schema_branch (str): The GitHub branch of the MDF schema to use in validation.
+                    See https://github.com/materials-data-facility/data-schemas
+        """
+        self.__dataset = None
+        self.__scroll_id = None
+        self.__ingest_date = datetime.utcnow().isoformat("T") + "Z"
+        self.__indexed_files = []
+        self.ref_resolver = jsonschema.RefResolver("https://raw.githubusercontent.com/materials-"
+                                                   "data-facility/data-schemas/{}/schemas/"
+                                                   .format(schema_branch), None)
+
+    def validate_mdf_dataset(self, ds_md, validation_info=None):
+        """Begin validating a new dataset against the MDF schema.
+
+        This function is a generator. You must initialize it with the following arguments,
+        and then use `.send()` to send in single records to validate.
+        The first yield value is the validated dataset entry.
+        When you .send() a record, the yield value is the validated record.
+        After you are finished sending records, you may send None to end the generator.
+        Every record intended for MDF must be validated through this method.
+
+        Arguments:
+            ds_md (dict): The dataset metadata to validate.
+            validation_info (dict): Additional validation configuration.
+                project_blocks (list of str): The allowed "project" blocks. Default None.
+                required_fields (list of str): Additional required fields not present
+                        in the MDF schema. Default None.
+                allowed_nulls (list of str): Fields allowed to be null/empty. Default None.
+                base_acl (list of str): The ACL to set on entries. Default None,
+                        which sets a public ACL.
+
+        Yields:
+            dict: Validated MDF-format metadata, ready for Search ingestion.
+                    The first yield value is always the validated dataset entry.
+                    When a record entry (dict) has been sent to the generator,
+                    this will be the validated record entry.
+
+        Accepts via .send():
+            dict: A record to validate.
+            None: Finish validating.
+        """
+        # Validate, save, and yield dataset
+        dataset = self._validate_dataset(ds_md, validation_info)
+        self.__dataset = dataset
+        # Fetch first record
+        record = yield dataset
+        # Process all records the user has
+        while record is not None:
+            record = yield self._validate_record(record)
+        # Yield once more to avoid forcing user to catch StopIteration
```

Comment:
This is pretty cool! I've never seen the `send` function before and I think it's a great use of it!